iwyu.cc

```diff
@@ -3543,6 +3572,12 @@ class InstantiatedTemplateVisitor
   // template-caller may or may not be responsible for.
   map<const Type*, const Type*> resugar_map_;
 
+  // When we see a full-use of a template argument we need to assign that full
+  // use to the template-caller.  Sometimes those uses are hidden behind
+  // type aliases (typedefs).  This maps those aliases to the underlying
+  // template arguments.
+  map<const Type*, const Type*> template_argument_aliases_;
```

Comment:
Does this need to be separate from the resugar map? I've tried a few times to get my head around the resugaring logic, without success, so I may be missing some subtlety, but they look closely related.

Reply:
Hmm, yeah, that sounds weird. I would expect the resugar map to do exactly what you're attempting here + maybe some  magic for dependent types.

Reply:
I haven't tested this idea yet, but I *think* that `resugar_map` is designed to help with situations where template arguments don't match with template parameters.  For example,

```
template<typename T>
struct array_traits;

template<typename T, size_t N>
struct array_traits<T[N]> {
  static size_t size = N;
};

int foo(int) {
  char my_string[] = "...";
  return array_traits<decltype(my_string)>::size;
}
```

Here when walking the specialization of `array_traits`, there are two template parameters (`T` and `N`) corresponding to a single template argument (`decltype(my_string)`) and we need some way to match them up.  I've certainly seen some code that appears to try to deal with this sort of situation, but I've not experimented with it myself.